README.md
# python-jedi package

Python Jedi based autocompletion plugin.

## Installation
Either use Atoms package manager or `apm install python-jedi`. Install autocomplete-plus before installing this package.

### Usage

python-jedi uses python3 interpreter in your path by default.

For python2 autocomplete go to settings -> check Enable Python2 and uncheck to use Python3.

#### To Use virtualenv/pyvenv 
  - Add virtualenv path or pyvenv path in the settings(Pathtopython field).(eg:/home/user/py3pyenv/bin/python3 or /home/user/py2virtualenv/bin/python).
  
  - Check 'Enable Pathtopython' option in package settings.
  
  - Make sure you are giving the correct virtualenv/pyvenv path. Otherwise autocomplete won't work. Also for python2 virtualenv path to work, you must have checked 'Enable Python2' option.

### Note

- If you are not using virtualenv path, please uncheck 'Enable Pathtopython'. 

- The completion daemon is started on port 7777 - please make sure no
other service is using this port.

- The completion daemon is stopped appropriately, which was fully tested in linux
environment. The package has not been tested under windows environment. It might
be buggy.

### Warning

Do not use it along with autocomplete-plus-jedi. Since it is originally forked
from it, python-jedi may malfunction. 


CHANGELOG.md
## 0.3.3
* Added type property to suggestion.

## 0.3.2
* Added goto_definitions functionality (issue #6).

## 0.3.1
* Fix for closing jedi-server in the background properly.

## 0.2.0
* Fix for issue #4

## 0.1.9
* Added new option 'enable pathtopython' to avoid logical confusion.

## 0.1.8
* Fixed Deprecated calls & added new dependency.

## 0.1.7
* No autocomplete disturbance after '#' comment.

## 0.1.6
* Python AutoComplete using jedi.


lib/jedi-python2-complete.py

import socket
import time
import json
import jedi
from BaseHTTPServer import BaseHTTPRequestHandler, HTTPServer


class http_completion(BaseHTTPRequestHandler):
    """
    Completion handler which returns the completions for a given source,
    line and cursor positon.
    """
    def _set_headers(self):
        """set the standard headers for a JSON response"""
        self.send_response(200)
        self.send_header("Content-type", "application/json")
        self.end_headers()

    def do_POST(self):
        """
        Payload to receive:
            source: whole source to parse
            line / column : current line and column

        Returns:
            array with dictionaries in it (name, description, docstring)
        """
        self._set_headers()

        length = int(self.headers.get('content-length', 0))
        read = self.rfile.read(length)

        read = json.loads(read)

        if read["type"] == "goto":
            payload = goto_def(read["source"], read["line"], read["column"], read["path"])
            payload = json.dumps(payload)
        else:
            payload = completions(read["source"], read["line"], read["column"], read["path"])
            payload = json.dumps(payload)

        self.wfile.write(payload)

        return


def run_server():
    """run the httpd"""
    address = ('', 7777)

    while True:
        try:
            print("Starting httpd")
            httpd = HTTPServer(address, http_completion)
            httpd.serve_forever()
        except (socket.error, KeyboardInterrupt) as exc:
            if exc.__class__ == KeyboardInterrupt:
                break

            # If we cannot bind to the port sleep wait 1 second and retry.
            # This can happen when reloading Atom e.x.
            time.sleep(1)


def completions(source, line, column, path):
    """
    generate list with completions for the line and column.

    Arguments:
        source: source code to generate completion for
        line, column: current cursor position in the source code

    Returns:
        list with dictionaries containing the name, docstring and description
        for all completions.
    """
    script = jedi.api.Script(
        source=source,
        line=line + 1,
        column=column,
        path = path
    )

    completions = list()

    try:
        for completion in script.completions():
            completions.append({
                "name": completion.name,
                "description": completion.description,
                "type":completion.type
            })

        return completions
    except:
        return []

def goto_def(source, line, column, path):

    try:
        script = jedi.api.Script(source, line, column, path)
        defs = script.goto_definitions()
    except:
        return []

    if defs:
        is_built = script.goto_definitions()[0].in_builtin_module()
        module_name = script.goto_definitions()[0].module_name

    defs_string = list()
    for get in defs:
        defs_string.append({
            "type": get.type,
            "description": get.description,
            "module_path": get.module_path,
            "line": get.line,
            "column":get.column,
            "is_built_in": is_built,
            "module_name": module_name
        })
        break

    return defs_string


if __name__ == "__main__":
    run_server()


lib/jedi-python3-complete.py
import json
from http.server import BaseHTTPRequestHandler, HTTPServer
import socket
import time

# sys.path.append(os.path.realpath(__file__))

import jedi


class http_completion(BaseHTTPRequestHandler):
    """
    Completion handler which returns the completions for a given source,
    line and cursor positon.
    """
    def _set_headers(self):
        """set the standard headers for a JSON response"""
        self.send_response(200)
        self.send_header("Content-type", "application/json")
        self.end_headers()

    def do_POST(self):
        """
        Payload to receive:
            source: whole source to parse
            line / column : current line and column

        Returns:
            array with dictionaries in it (name, description, docstring)
        """
        self._set_headers()

        content_length = self.headers.get('Content-Length')
        length = int(content_length)
        read = self.rfile.read(length).decode('utf-8')
        read = json.loads(read)

        if read["type"] == "goto":
            payload = goto_def(read["source"], read["line"], read["column"], read["path"])
            payload = json.dumps(payload)

        else:
            payload = completions(read["source"], read["line"], read["column"], read["path"])
            payload = json.dumps(payload)

        self.wfile.write(bytes(payload,'utf-8'))
        return


def run_server():
    """run the httpd"""
    address = ('', 7777)

    while True:
        try:
            print("Starting httpd")
            httpd = HTTPServer(address, http_completion)
            httpd.serve_forever()
        except (socket.error, KeyboardInterrupt) as exc:
            if exc.__class__ == KeyboardInterrupt:
                break

            # If we cannot bind to the port sleep wait 1 second and retry.
            # This can happen when reloading Atom e.x.
            time.sleep(1)


def completions(source, line, column, path):
    """
    generate list with completions for the line and column.

    Arguments:
        source: source code to generate completion for
        line, column: current cursor position in the source code

    Returns:
        list with dictionaries containing the name, docstring and description
        for all completions.
    """
    script = jedi.api.Script(
        source = source,
        line = line + 1,
        column = column,
        path = path
    )

    completions = list()

    try:
        for completion in script.completions():
            completions.append({
                "name": completion.name,
                "description": completion.description,
                "type":completion.type
            })

        return completions
    except:
        return []

def goto_def(source, line, column, path):

    try:
        script = jedi.api.Script( source, line , column ,path)
        defs = script.goto_definitions()
    except:
        return []

    if defs:
        is_built = script.goto_definitions()[0].in_builtin_module()
        module_name = script.goto_definitions()[0].module_name

    defs_string = list()
    for get in defs:
        defs_string.append({
            "type": get.type,
            "description": get.description,
            "module_path": get.module_path,
            "line": get.line,
            "column":get.column,
            "is_built_in": is_built,
            "module_name": module_name
        })
        break

    return defs_string

if __name__ == "__main__":
    run_server()
